rust/src/io/reader.rs

```diff
@@ -459,87 +462,115 @@ async fn read_struct_array(
     Ok(Arc::new(StructArray::from(sub_arrays)))
 }
 
-async fn read_list_array(
+async fn take_list_array<T: ArrowNumericType>(
     reader: &FileReader<'_>,
     field: &Field,
     batch_id: i32,
-    params: &ReadBatchParams,
-) -> Result<ArrayRef> {
-    let page_info = get_page_info(&reader.page_table, field, batch_id)?;
+    positions: &PrimitiveArray<T>,
+    indices: &UInt32Array,
+) -> Result<ArrayRef>
+where
+    T::Native: ArrowNativeTypeOp + OffsetSizeTrait,
+{
+    let first_idx = indices.value(0);
+    // Range of values for each index
+    let ranges = indices
+        .values()
+        .iter()
+        .map(|i| (*i - first_idx).as_usize())
+        .map(|idx| positions.value(idx).as_usize()..positions.value(idx + 1).as_usize())
```

Comment:
what about the last position? or does positions have a last value outside the indices ?

Reply:
yes, so this range is `exclusive`, and offsets array has 1 more element than the value array.

The `last` element is tested in the unit test as well.